usaspending_api/awards/v2/filters/award.py

```diff
@@ -0,0 +1,204 @@
+            or_queryset = Award.objects.none()
+            for v in value:
+                if v["lower_bound"] is not None and v["upper_bound"] is not None:
+                    or_queryset |= Award.filter(total_obligation__gt=v["lower_bound"],
+                                                total_obligation__lt=v["upper_bound"])
+                elif v["lower_bound"] is not None:
+                    or_queryset |= Award.filter(total_obligation__gt=v["lower_bound"])
+                elif v["upper_bound"] is not None:
+                    or_queryset |= Award.filter(total_obligation__lt=v["upper_bound"])
+                else:
+                    raise InvalidParameterException('Invalid filter: award amount has incorrect object.')
+            queryset &= or_queryset
+
+        # award_ids - DONE
+        elif key == "award_ids":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= Award.filter(id=v)
+            queryset &= or_queryset
+
+        # program_numbers  - DONE
+        elif key == "program_numbers":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= Award.fitler(
+                    transaction__assistance_data__cfda__program_number=v)
+            queryset &= or_queryset
+
+        # naics_codes - DONE
+        elif key == "naics_codes":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= award_filter().fitler(
+                    transaction__contract_data__naics=v)
+            queryset &= or_queryset
+
+        # psc_codes - DONE
+        elif key == "psc_codes":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= Award.fitler(
+                    transaction__contract_data__product_or_service_code=v)
+            queryset &= or_queryset
+
+        # contract_pricing_type_codes - DONE
+        elif key == "contract_pricing_type_codes":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= Award.fitler(
+                    transaction__contract_data__type_of_contract_pricing=v)
+            queryset &= or_queryset
+        # set_aside_type_codes - DONE
+        elif key == "set_aside_type_codes":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= Award.fitler(
+                    transaction__contract_data__type_set_aside=v)
+            queryset &= or_queryset
+        # extent_competed_type_codes - DONE
+        elif key == "extent_competed_type_codes":
+            or_queryset = Award.objects.none()
+            for v in value:
+                or_queryset |= Award.fitler(
+                    transaction__contract_data__extent_competed=v)
+            queryset &= or_queryset
+
+        else:
+            raise InvalidParameterException('Invalid filter: ' + key + ' does not exist.')
```

Comment:
I think we can make this check a bit more efficient. Right now it goes through and checks if it matches any of the previous keys, if not _then_ raise the exception. We can do something like `if key not in list_of_valid_keys` and raise the exception earlier. Just a thought on performance. Not a major impact but if more and more filter options are added, it can take time to error out on a faulty key.

Reply:
While it would look nicer, that would only slow down the code.
for success cases (key is in list_of_valid_keys): it would slow everything down by~ list_of_valid_keys/2 because it would iterate through the list_of_valid_keys and then the if, elif statements

for failure cases it would offer no speedup (it would simply go through the elifs and fail which takes exactly the same amount of time as iterating through list_of_valid_keys and failing)

Reply:
Lookups in a dictionary are O(1) and would be the fastest for detecting a failing key

Reply:
ah list_of_valid_keys is actually dict_of_valid_keys.  Thats fine then